### PR TITLE
refactor: Refactor `Reduce` operation.

### LINKED ISSUE #256: Reduction operations should return a single value
## Context

As described [here](https://github.com/loophp/collection/pull/255#discussion_r895220913), user experience would be improved if reduction operations would return a value directly, rather than a collection.

While in some cases it might be useful to have these operations return a collection, it's much more common for users to expect these to return a single value; after all, that's the typical behaviour of reductions. This is the case in other languages (just a couple examples):

- [Java Streams](https://www.logicbig.com/tutorials/core-java-tutorial/java-util-stream/reduction.html)
- [Rust std::iter::Iterator::fold](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.fold)

This change would make Collection easier to use and also make it easier to differentiate between some operations, like [Reduce](https://loophp-collection.readthedocs.io/en/latest/pages/api.html#reduce) (will return a single value) and [Reduction](https://loophp-collection.readthedocs.io/en/latest/pages/api.html#reduction) (will continue to return a collection).

## Steps required to reproduce the problem

```PHP
$result = Collection::fromIterable([1, 4, 3, 0, 2])->min();
```

## Expected Result

```PHP
// 0
```

## Actual Result

```PHP
// Collection([4 => 4])
```

## Operations
- compare, min, max
- foldLeft, foldLeft1
- foldRight, foldRight1
- reduce

### PR DESCRIPTION
**Pull request by drupol**
BREAKING CHANGE: yes

This PR:

* [x] Fix #256 
* [x] It breaks backward compatibility
* [x] Is covered by unit tests
* [x] Has static analysis tests (psalm, phpstan)
* [x] Has documentation
* [ ] Is an experimental thing

**DONE** 🟢 

- [x] first
- [x] foldLeft
- [x] foldLeft1
- [x] foldRight
- [x] foldRight1
- [x] implode
- [x] reduce
- [x] unlines
- [x] unwords
- [x] last
- [x] head
- [x] get
- [x] compare
- [x] min
- [x] max

### CODE REVIEW

**Comment by drupol** [comment]
I think I'm done with this PR, let's leave it open a little bit to carefully check if this is the proper thing to do.

I know the pace is a bit different during holidays, so, it might take a little while before it gets merged.

**Comment by drupol** [comment]
@AlexandruGG I think it's ready, wdyt ?

**Comment by AlexandruGG** [comment]
> @AlexandruGG I think it's ready, wdyt ?

Amazing! I'll go through this today :)

File: src/Contract/Operation/Firstable.php
Comment (AlexandruGG) [inline src/Contract/Operation/Firstable.php:27]:
```suggestion
     * @return T|null
```

`first` can return _null_

**Review by AlexandruGG — COMMENTED** [review]
Nice work on this, I know it's a lot!

Here's the status from what I can see:

**DONE** 🟢 

- [x] first
- [x] foldLeft
- [x] foldLeft1
- [x] foldRight
- [x] foldRight1
- [x] implode
- [x] reduce
- [x] unlines
- [x] unwords


**MISSING** 🟡 

- [x] last
- [ ] head
- [ ] get
- [ ] compare
- [ ] min
- [ ] max

**Comment by AlexandruGG** [comment]
> @AlexandruGG I think it's ready, wdyt ?

@drupol let me know if you'd like me to help with the remaining ones :)

**Comment by drupol** [comment]
Thanks for the review Alex, I can continue to work on it tonight, but feel free to push things if you want to.

**Comment by AlexandruGG** [comment]
> Thanks for the review Alex, I can continue to work on it tonight, but feel free to push things if you want to.

Sure, leave the comparison operations to me :)

File: src/Contract/Operation/Firstable.php
Comment (drupol) [reply src/Contract/Operation/Firstable.php]: Fixed, fixuped, rebased.

**Comment by drupol** [comment]
I now leave this PR to you @AlexandruGG , good work !

**Comment by AlexandruGG** [comment]
> I now leave this PR to you @AlexandruGG , good work !

Thanks! I added the remaining operations and updated a couple bits of documentation that were still mentioning `current`.

I'll give this another fresh look tomorrow, see if I can spot anything else needed.

**Comment by drupol** [comment]
> Thanks! I added the remaining operations and updated a couple bits of documentation that were still mentioning `current`.

Indeed, nice. I was planning to do it after all the commits were done, but you did it. Thanks :)
 
> I'll give this another fresh look tomorrow, see if I can spot anything else needed.

Ok will do that ! Cheers!

**Comment by AlexandruGG** [comment]
@drupol I went through this again and added a few more things. Please feel free to comb through the docs yourself as well and see if you can spot any bits we might have missed

**Review by AlexandruGG — APPROVED** [review]
Thank you for working on this! 💯

**Comment by drupol** [comment]
This is super great !!!

I propose to merge this, then do remaining changes if any in another branch.

We've been waiting for too long for this :)

**Pull request merged**
